### PR TITLE
i/5828: Remove entire exception when collapsed selection is inside text with restricted attribute

### PR DESCRIPTION
**Pull request by panr**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Remove entire exception when collapsed selection is inside text with restricted attribute. Closes https://github.com/ckeditor/ckeditor5/issues/5828.

---

### Additional information

Is it **breaking change**? 🤔

### CODE REVIEW

File: src/restrictededitingexceptioncommand.js [hunk 1]
```diff
@@ -38,10 +38,32 @@ export default class RestrictedEditingExceptionCommand extends Command {
 			const ranges = model.schema.getValidRanges( selection.getRanges(), 'restrictedEditingException' );
 
 			if ( selection.isCollapsed ) {
-				if ( valueToSet ) {
-					writer.setSelectionAttribute( 'restrictedEditingException', true );
-				} else {
+				const position = selection.getFirstPosition();
+
+				// When selection is inside a restricted text
+				if ( selection.hasAttribute( 'restrictedEditingException' ) ) {
```
Comment (jodator) [inline src/restrictededitingexceptioncommand.js]: After all the changes this check is not required - basically, you need to check the `valueToSet` first (comment is somwhere below).

File: src/restrictededitingexceptioncommand.js [hunk 2]
```diff
@@ -38,10 +38,32 @@ export default class RestrictedEditingExceptionCommand extends Command {
...
 					writer.removeSelectionAttribute( 'restrictedEditingException' );
+
+					// Remove entire exception if the caret isn't at the end of a restricted text
+					if ( selection.anchor.textNode !== null ) {
+						writer.removeAttribute( 'restrictedEditingException', restrictedEditingExceptionRange );
+					}
+				} else if ( valueToSet ) {
+					// Set attribute on selection with unset attribute
+					writer.setSelectionAttribute( 'restrictedEditingException', valueToSet );
 				}
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/restrictededitingexceptioncommand.js]: After checking how this works with the `valueToSet` the above `if` should be reversed.

```js
writer.removeSelectionAttribute( 'restrictedEditingException' );

const isSameException = value => value.item.getAttribute( 'restrictedEditingException' ) === this.value;
const focus = selection.focus;

const exceptionStart = focus.getLastMatchingPosition( isSameException, { direction: 'backward' } );
const exceptionEnd = focus.getLastMatchingPosition( isSameException );

if ( !( focus.isEqual( exceptionStart ) || focus.isEqual( exceptionEnd ) ) ) {
	writer.removeAttribute( 'restrictedEditingException', writer.createRange( exceptionStart, exceptionEnd ) );
}
```

oor something like this.

File: tests/restrictededitingexceptioncommand.js [hunk 3]
```diff
@@ -117,12 +117,57 @@ describe( 'RestrictedEditingExceptionCommand', () => {
 				expect( model.document.selection.hasAttribute( 'restrictedEditingException' ) ).to.be.true;
 			} );
 
-			it( 'should remove the selection attribute if a text has the non-restricted attribute', () => {
+			it( 'should remove attribute from text node if a text has the non-restricted attribute', () => {
```
Comment (jodator) [inline tests/restrictededitingexceptioncommand.js]:
```suggestion
			it( 'should remove an attribute from text node if a text has the non-restricted attribute', () => {
```

**Comment by panr** [comment]
I hope I fixed all issues 😉 Ready to re-review.

**Pull request merged**